### PR TITLE
[OCPCLOUD-1662] Implement Control Plane Machine Set manifest generator

### PR DESCRIPTION
**Pull request by damdo**
This PR introduces a cli tool which once given access to an OCP cluster is able to generate the canonical Control Plane Machine Set manifest for that cluster, which cluster admins can apply to start managing the control plane of their clusters through the Control Plane Machine Set Operator.

Env test base testing will follow in a follow up PR.
Docs will follow in a follow up PR.

### CODE REVIEW

File: pkg/cmd/generate/generator.go [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+	machinev1 "github.com/openshift/api/machine/v1"
+	machinev1beta1 "github.com/openshift/api/machine/v1beta1"
+	configclient "github.com/openshift/client-go/config/clientset/versioned"
+	machinev1builder "github.com/openshift/client-go/machine/applyconfigurations/machine/v1"
+	machinev1beta1builder "github.com/openshift/client-go/machine/applyconfigurations/machine/v1beta1"
+	machineclient "github.com/openshift/client-go/machine/clientset/versioned"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type options struct {
```
Comment (JoelSpeed) [inline pkg/cmd/generate/generator.go:41]: Comment?

File: pkg/cmd/generate/generator.go [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+	Kubeconfig string
+}
+
+var (
+	// generateOpts captures the options that affect cpms generation.
+	generateOpts = options{}
+)
+
+const (
+	namespace                      = "openshift-machine-api"
```
Comment (JoelSpeed) [inline pkg/cmd/generate/generator.go:52]:
```suggestion
	machineAPINamespace = "openshift-machine-api"
```

File: pkg/cmd/generate/generator.go [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+	label                          = "machine.openshift.io/cluster-api-machine-role=master"
```
Comment (JoelSpeed) [inline pkg/cmd/generate/generator.go:53]: Maybe?
```suggestion
	masterMachineSelector  = "machine.openshift.io/cluster-api-machine-role=master"
```

File: pkg/cmd/generate/generator.go [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+// NewGenerateCmd implements the "generate" subcommand for cpms manifest generation.
+func NewGenerateCmd() *cobra.Command {
+	generateCmd := &cobra.Command{
+		Use:   "generate",
+		Short: "Manage generation of the Control Plane Machine Set manifest",
+		Long:  "Generating Control Plane Machine Set manifest",
+		Run:   generateCmdFunc,
+	}
+
+	generateCmd.PersistentFlags().StringVar(&generateOpts.To, "to", "", "User-defined destination file for the generated manifest")
```
Comment (JoelSpeed) [inline pkg/cmd/generate/generator.go:71]: When omitted prints to stdout?

File: pkg/cmd/generate/generator.go [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+	generateCmd.PersistentFlags().StringVar(&generateOpts.Kubeconfig, "kubeconfig", "", "User-defined kubeconfig path for the cluster")
```
Comment (JoelSpeed) [inline pkg/cmd/generate/generator.go:72]: When omitted ...?

File: pkg/cmd/generate/generator.go [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+
+// generateCmdFuc specifies a function to handle the generate command.
+func generateCmdFunc(cmd *cobra.Command, args []string) {
+	generatedCPMS, err := generateCPMSManifest()
+	if err != nil {
+		log.Print(err)
+		os.Exit(1)
+	}
+
+	fmt.Print(generatedCPMS)
```
Comment (JoelSpeed) [inline pkg/cmd/generate/generator.go:85]: Print the manifest always?

File: pkg/cmd/generate/generator.go [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+	}
+
+	cpms := machinev1builder.ControlPlaneMachineSet(cpmsObjectName, namespace).WithSpec(&cpmsSpec)
+
+	yamlCpms, err := yaml.Marshal(cpms)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal cpms yaml: %w", err)
+	}
+
+	return string(yamlCpms), nil
```
Comment (JoelSpeed) [inline pkg/cmd/generate/generator.go:119]: Does this need to be a string? I'd probably expect a `[]byte` to be returned and then written to stdout or to a file

File: pkg/cmd/generate/generator.go [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+}
+
+// generateCPMSSpec generates a CPMS spec with provider specific details by detecting the provider and mapping its failure domains.
+func generateCPMSSpec(ctx context.Context, machineClientSet *machineclient.Clientset, configClientSet *configclient.Clientset) (machinev1builder.ControlPlaneMachineSetSpecApplyConfiguration, error) {
+	machineSetClient := machineClientSet.MachineV1beta1().MachineSets(namespace)
+	machineClient := machineClientSet.MachineV1beta1().Machines(namespace)
+
+	msetList, err := machineSetClient.List(ctx, metav1.ListOptions{})
```
Comment (JoelSpeed) [inline pkg/cmd/generate/generator.go:127]: Nit: I would probably expect this to be `machineSets` and the one below to be `machines`

File: pkg/cmd/generate/generator.go [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+
+	var spec machinev1builder.ControlPlaneMachineSetSpecApplyConfiguration
+
+	switch infra.Spec.PlatformSpec.Type {
+	case configv1.AWSPlatformType:
+		spec, err = generateCPMSAWSSpec(msetList, mList)
+		if err != nil {
+			return machinev1builder.ControlPlaneMachineSetSpecApplyConfiguration{}, fmt.Errorf("error listing Machines: %w", err)
+		}
+	case configv1.AzurePlatformType:
```
Comment (JoelSpeed) [inline pkg/cmd/generate/generator.go:163]: Maybe make this the default case for now and we can add other platforms as we go?

File: pkg/cmd/generate/generator.go [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+
+	// Decode the ProviderSpec for modification.
+	ps, err := providerSpecFromRawExtension(mList.Items[0].Spec.ProviderSpec.Value)
+	if err != nil {
+		return machinev1builder.ControlPlaneMachineSetSpecApplyConfiguration{}, fmt.Errorf("failed to decode machine providerSpec: %w", err)
+	}
+
+	// Remove Failure domains already present in the Control Plane Machine Set spec.
+	ps.Subnet = machinev1beta1.AWSResourceReference{}
+	ps.Placement.AvailabilityZone = ""
```
Comment (JoelSpeed) [inline pkg/cmd/generate/generator.go:183]: Are you clearing the subnet too?

File: pkg/cmd/generate/generator.go (hunk 4 above)
Comment (damdo) [reply pkg/cmd/generate/generator.go]: Yes, forgot to check it in, done.

File: pkg/cmd/generate/generator.go (hunk 10 above)
Comment (damdo) [reply pkg/cmd/generate/generator.go:199]: Yes that should happen in the line above.

**Comment by JoelSpeed** [comment]
Changes LGTM, would like to squash before merging

**Comment by JoelSpeed** [comment]
/override ci/prow/e2e-aws-serial 

This code change doesn't actually affect the payload, and the serial tests seem to be broken

**Pull request merged**